tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/llm/TestParallelFFNLayer.java

```diff
@@ -0,0 +1,420 @@
+
+            // Step 2: RMSNorm sequence
+            .task("reduce", TestParallelFFNLayer::reduceSquareSums, context, x, reduce, localSize)
+            .task("sum", TestParallelFFNLayer::finalSum, context, reduce, dim, rmsNormEps)
+            .task("ns", TestParallelFFNLayer::normalizeAndScale2, context, temp, x, rmsWeight, reduce, dim)
+
+            // Step 3: Parallel projections with W1 and W3
+            .task("projection1", TestParallelFFNLayer::matrixVectorMultiply, context, temp, hidden, w1, dim, hiddenDim)
+            .task("projection3", TestParallelFFNLayer::matrixVectorMultiply, context, temp, hidden2, w3, dim, hiddenDim)
+
+            // Step 4: SiLU activation and element-wise multiplication
+            .task("silu", TestParallelFFNLayer::siluActivation, context, hidden)
+            .task("multiply", TestParallelFFNLayer::elementMultiply, context, hidden2, hidden)
+
+            // Step 5: Final projection and residual
+            .task("projection2", TestParallelFFNLayer::matrixVectorMultiply, context, hidden, temp, w2, hiddenDim, dim)
+            .task("residual2", TestParallelFFNLayer::addInPlace, context, temp, x)
+
+            // Transfer result back to host
+            .transferToHost(DataTransferMode.EVERY_EXECUTION, x);
+        // @formatter:on
+
+        // Create worker grids for different task dimensions
+        GridScheduler gridScheduler = getGridScheduler(dim, localSize, hiddenDim);
+
+        // Create execution plan
+        ImmutableTaskGraph immutableTaskGraph = taskGraph.snapshot();
+
+        // Execute the task graph
+        System.out.println("Executing parallel FFN layer...");
+        try (TornadoExecutionPlan executionPlan = new TornadoExecutionPlan(immutableTaskGraph)) {
+            executionPlan.withGridScheduler(gridScheduler).execute();
+            System.out.println("Parallel execution completed successfully");
+        } catch (Exception e) {
+            System.out.println("Exception during execution: " + e.getMessage());
+            e.printStackTrace();
+        }
+
+        // Verify results
+        if (DEBUG) {
+            System.out.println("\nComparing results:");
+            boolean allMatch = true;
+            for (int i = 0; i < dim; i++) {
+                float expected = xSeq.get(i);
+                float actual = x.get(i);
+                float relError = Math.abs(expected - actual) / (Math.abs(expected) + 1e-6f);
+
+                System.out.printf("Element %2d: Expected=%.7f, Actual=%.7f, RelError=%.7f%n", i, expected, actual, relError);
+
+                if (relError > 0.01f) {
+                    allMatch = false;
+                }
+            }
+
+            if (allMatch) {
+                System.out.println("All values match within tolerance!");
+            } else {
+                System.out.println("Some values don't match within tolerance.");
+            }
+        }
```

Comment:
Similar comment with previous tests. Can we replace the print and conditional checks with the invocation of the `assertEquals` and add the error fraction for the acceptable deviation due to the precision difference?